capirca/utils/iputils.py

```diff
@@ -0,0 +1,52 @@
+from ipaddress import summarize_address_range, _BaseNetwork
+
+
+def exclude_address(base_net: _BaseNetwork, exclude_net: _BaseNetwork):
+  '''
+  Function to exclude a subnetwork from another, returning a generator that yields
+  all values that correspond to the base network without the exclude network.
+
+  This is functionally equivalent to the _BaseNetwork "address_exclude" from the
+  `ipaddress` standard library, but is a faster implementation since
+  the standard library function is a O(n) operation on the length of the
+  netmask of the excluding network, whereas this function is O(1) for all cases.
+
+  args:
+    base_net: an object of type _BaseNetwork, the network that
+              contains the exclude network
+    exclude_net: an object of type _BaseNetwork, the network
+                 that is being removed from the base_net
+  raises:
+    ValueError if exclude_net is not completely contained in base_net
+  '''
+
+  if not base_net._version == exclude_net._version:
+    raise TypeError("%s and %s are not of the same version" % (
+      base_net, exclude_net))
+
+  if not isinstance(exclude_net, _BaseNetwork):
```

Comment:
Should we check this first, before we access any members? Do we need to check the type of base_net as well?

Reply:
True! With the stdlib code it's not necessary since it's being called from the class but with this one that's a good extra check